networkit/cpp/centrality/TopCloseness.cpp

```diff
@@ -1,649 +1,626 @@
-        reachU_without_max_scc[V] += reachU_without_max_scc[W];
-      }
-      reachU_scc[V] += reachU_scc[W];
-      reachU_scc[V] = std::min(reachU_scc[V], n);
-      reaches_max_scc[V] = reaches_max_scc[V] || reaches_max_scc[W];
-    });
+    // Dynamic programming to compute number of reachable vertices
+    for (count V = 0; V < N; V++) {
+        if (V == maxSizeCC) {
+            continue;
+        }
+        sccGraph.forNeighborsOf(V, [&](node W) {
+            reachL_scc[V] = std::max(reachL_scc[V], reachL_scc[W]);
+            if (!reach_from_max_scc[W]) {
+                reachU_without_max_scc[V] += reachU_without_max_scc[W];
+            }
+            reachU_scc[V] += reachU_scc[W];
+            reachU_scc[V] = std::min(reachU_scc[V], n);
+            reaches_max_scc[V] = reaches_max_scc[V] || reaches_max_scc[W];
+        });
 
-    if (reaches_max_scc[V]) {
-      reachU_scc[V] = reachU_without_max_scc[V] + reachU_scc[V];
+        if (reaches_max_scc[V]) {
+            reachU_scc[V] = reachU_without_max_scc[V] + reachU_scc[V];
+        }
+        reachL_scc[V] += sccs_vec[V].size();
+        reachU_scc[V] += sccs_vec[V].size();
+        reachU_scc[V] = std::min(reachU_scc[V], n);
     }
-    reachL_scc[V] += sccs_vec[V].size();
-    reachU_scc[V] += sccs_vec[V].size();
-    reachU_scc[V] = std::min(reachU_scc[V], n);
-  }
-
-  for (count v = 0; v < n; v++) {
-    reachL[v] = reachL_scc[sccs.componentOfNode(v) - 1];
-    reachU[v] = reachU_scc[sccs.componentOfNode(v) - 1];
-    if (false) { // MICHELE: used to check if the bounds are correct
-      count r = 0;
-      Traversal::BFSfrom(G, v, [&](node, count) { r++; });
-
-      if (reachL[v] > r || reachU[v] < r) {
-        DEBUG("BIG MISTAKE! ", reachL[v], " ", r, " ", reachU[v]);
-        while (true) {
+
+    for (count v = 0; v < n; v++) {
+        reachL[v] = reachL_scc[sccs.componentOfNode(v) - 1];
+        reachU[v] = reachU_scc[sccs.componentOfNode(v) - 1];
+        if (false) { // MICHELE: used to check if the bounds are correct
+            count r = 0;
+            Traversal::BFSfrom(G, v, [&](node, count) { r++; });
+
+            if (reachL[v] > r || reachU[v] < r) {
+                DEBUG("BIG MISTAKE! ", reachL[v], " ", r, " ", reachU[v]);
+                while (true) {
+                }
+            }
         }
-      }
     }
-  }
 }
 
 void TopCloseness::computeReachableNodesUndir() {
-  reachL = std::vector<count>(n);
-
-  ConnectedComponents comps(G);
-  comps.run();
-  std::map<index, count> sizes = comps.getComponentSizes();
-  G.forNodes([&](node v) {
-    index cv = comps.componentOfNode(v);
-    reachL[v] = sizes[cv];
-  });
-  reachU = reachL;
+    reachL = std::vector<count>(n);
+
+    ConnectedComponents comps(G);
+    comps.run();
+    std::map<index, count> sizes = comps.getComponentSizes();
+    G.forNodes([&](node v) {
+        index cv = comps.componentOfNode(v);
+        reachL[v] = sizes[cv];
+    });
+    reachU = reachL;
```

Comment:
Don't know how much could be saved here in terms of memory, but since `reachU` and `reachL` are the same for undirected graphs, maybe it would be a good idea to make these vectors references/pointers?

Reply:
That's a good point, I'll address that.